### PR TITLE
Migrate to Java 16 records (part 2)

### PR DESCRIPTION
**Pull request by arteam**
Try to represent immutable data with Java records introduced in [JEP 395](https://openjdk.java.net/jeps/395)

Convert only existing immutable classes, no "POJO with setters to a record" refactorings.

### CODE REVIEW

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDocTests.java [hunk 1]
```diff
@@ -351,9 +351,7 @@ public void testToXContent() throws IOException {
         when(mockJvmInfo.getBundledJdk()).thenReturn(true);
         when(mockJvmInfo.getUsingBundledJdk()).thenReturn(true);
 
-        final Build mockBuild = mock(Build.class);
-        when(mockBuild.flavor()).thenReturn(Build.Flavor.DEFAULT);
-        when(mockBuild.type()).thenReturn(Build.Type.DOCKER);
+        final Build mockBuild = new Build(Build.Flavor.DEFAULT, Build.Type.DOCKER, "", "", false, "");
```
Comment (grcevski) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDocTests.java:354]: Similar question about this test change.

**Review by grcevski — APPROVED** [review]
LGTM! Thanks Artem.

I left two questions about some small test changes.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/SegmentCountStepTests.java [hunk 2]
```diff
@@ -134,10 +134,10 @@ public void testIsConditionIsTrueEvenWhenMoreSegments() {
         Index index = new Index(randomAlphaOfLengthBetween(1, 20), randomAlphaOfLengthBetween(1, 20));
         IndicesSegmentResponse indicesSegmentResponse = Mockito.mock(IndicesSegmentResponse.class);
         IndexSegments indexSegments = Mockito.mock(IndexSegments.class);
-        IndexShardSegments indexShardSegments = Mockito.mock(IndexShardSegments.class);
-        Map<Integer, IndexShardSegments> indexShards = Collections.singletonMap(0, indexShardSegments);
         ShardSegments shardSegmentsOne = Mockito.mock(ShardSegments.class);
         ShardSegments[] shardSegmentsArray = new ShardSegments[] { shardSegmentsOne };
+        IndexShardSegments indexShardSegments = new IndexShardSegments(ShardId.fromString("[idx][123]"), shardSegmentsArray);
```
[the remark this answers is not in the document]
Comment (arteam) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/SegmentCountStepTests.java:139]: It used to be a mock via `Mockito.mock(IndexShardSegments.class)`, but with the change `IndexShardSegments` is now a record which is `final`. Mockito can't mock final classes without additional Java agent instrumentation. So we have to use a "real" data class here.

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDocTests.java (hunk 1 above)
Comment (arteam) [reply x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDocTests.java:354]: Same purpose here. We can't mock records and Mockito usually advice against mocking data classes.

**Comment by arteam** [comment]
Thanks Nikola!

**Pull request merged**